gitcache/main.go

```diff
@@ -4,139 +4,191 @@ import (
 	if err != nil {
-		log.Panic(err)
+		sugar.Panic(err)
 	}
 
-	defer os.RemoveAll(dir)
-	defer os.RemoveAll(tarDir)
+	defer os.RemoveAll(storage.GitDir)
+	defer os.RemoveAll(storage.GitTarDir)
+	defer os.RemoveAll(storage.BlobArchiveDir)
+	defer os.RemoveAll(gittar)
 
-	// Clone the given repository to the given directory
-	r, err := git.PlainClone(dir, false, &git.CloneOptions{
-		URL:   fmt.Sprintf("http://%s/%s/%s", repo.Host, repo.Owner, repo.Repo),
-		Depth: 1, // Just fetch the last commit
-	})
+	// checks if there is an existing git repo in the bucket
+	data, exists := bucket.Get(storage.BlobGitFolderPath)
+	alreadyUptoDate := false
+
+	if exists {
+		sugar.Info("bucket ", blob, " already has git folder")
+		gitRepo, alreadyUptoDate, err = fetchGitRepo(&storage, data)
+	} else {
+		sugar.Info("bucket ", blob, " does not have a git folder")
+		gitRepo, err = cloneGitRepo(&storage, repo)
+	}
 	if err != nil {
-		log.Panic(err)
+		sugar.Panic(err)
+	}
+
+	// case where the git repository hasn't changed and it need not be updated into the blob
+	if alreadyUptoDate {
+		sugar.Info("bucketb ", blob, " git folder is already up to date.")
+		// Just update the last sync time and return
+		err = bucket.Set(storage.BlobLastSyncPath, []byte(fmt.Sprintf("%64b", time.Now().Unix())))
+		if err != nil {
+			sugar.Panic(err)
+		}
+		sugar.Info("Finished processing ", repo)
+		return
 	}
-	lastRef, err := r.Head()
+
+	lastRef, err := gitRepo.Head()
 	if err != nil {
-		log.Panic(err)
+		sugar.Panic(err)
 	}
 
-	commit, err := r.CommitObject(lastRef.Hash())
+	commit, err := gitRepo.CommitObject(lastRef.Hash())
 	if err != nil {
-		log.Panic(err)
+		sugar.Panic(err)
 	}
 
-	// opening the blob
-	bucket, err := New(blob)
+	sugar.Info("Storing the last commit ", commit.Author.When)
+	// storing the last commit to the blob
+	err = bucket.Set(storage.BlobLastCommitPath, []byte(fmt.Sprintf("%64b", commit.Author.When.Unix())))
 	if err != nil {
 		log.Panic(err)
 	}
 
-	archivePath := path.Join(tarDir, fmt.Sprintf("%s.tar.gz", repo.Repo))
-	// storing the last commit to the blob
-	err = bucket.Set(fmt.Sprintf("gitcache/%s/%s/lastcommit", repo.Owner, repo.Repo), []byte(fmt.Sprintf("%64b\n", commit.Author.When.Unix())))
+	data, err = archiveFolder(storage.GitDir, storage.GitTarFile)
 	if err != nil {
 		log.Panic(err)
 	}
 
-	now := time.Now().Unix()
-	err = bucket.Set(fmt.Sprintf("gitcache/%s/%s/lastsync", repo.Owner, repo.Repo), []byte(fmt.Sprintf("%64b\n", now)))
+	err = bucket.Set(storage.BlobGitFolderPath, data)
 	if err != nil {
 		log.Panic(err)
 	}
 
 	// removing the .git folder as it is not required for the tar ball
-	err = os.RemoveAll(path.Join(dir, ".git"))
+	err = os.RemoveAll(path.Join(storage.GitDir, ".git"))
 	if err != nil {
 		log.Panic(err)
 	}
 
-	// creating an archive
-	archive := archiver.NewTarGz()
-
-	err = archive.Archive([]string{dir}, archivePath)
+	data, err = archiveFolder(storage.GitDir, storage.BlobArchiveFile)
```

Comment:
If this errors out for some reason, the bucket will be left in a bad state, as we already stored the last commit information etc prior to this. 

Let's write the metadata *after* we archive and upload successfully? 

Have you also considered using GCS metadata instead of individual files as well, which would also prevent these kind of de-sync issues?

Reply:
> If this errors out for some reason, the bucket will be left in a bad state, as we already stored the last commit information etc prior to this.

> Let's write the metadata after we archive and upload successfully?

Agreed, moved it to the end of the upload. 

> Have you also considered using GCS metadata instead of individual files as well, which would also prevent these kind of de-sync issues?

I wasn't aware of that. I agree it is a valid point. In the upcoming PR's we can fix this.